pytket/extensions/cutensornet/tnstate/ttn.py

```diff
@@ -0,0 +1,845 @@
+            A ``libhandle`` should be created via a ``with CuTensorNet() as libhandle:``
+            statement. The device where the TTN is stored will match the one specified
+            by the library handle.
+
+            The current implementation requires the keys of ``qubit_partition`` to be
+            integers from ``0`` to ``2^l - 1`` for some ``l``.
+
+        Args:
+            libhandle: The cuTensorNet library handle that will be used to carry out
+                tensor operations on the TTN.
+            qubit_partition: A partition of the qubits in the circuit into disjoint
+                groups, describing the hierarchical structure of the TTN. As a rule of
+                thumb, the cost of a gate between qubits on ``qubit_partition[i]`` and
+                ``qubit_partition[j]`` scales exponentially on ``i XOR j``.
+            config: The object describing the configuration for simulation.
+
+        Raises:
+            ValueError: If the keys of ``qubit_partition`` do not range from ``0`` to
+                ``2^l - 1`` for some ``l``.
+            ValueError: If a ``Qubit`` is repeated in ``qubit_partition``.
+            ValueError: If there is only one entry in ``qubit_partition``.
+            NotImplementedError: If the value of ``truncation_fidelity`` in ``config``
+                is smaller than one.
+        """
+        self._lib = libhandle
+        self._cfg = config
+        self._logger = set_logger("TTN", level=config.loglevel)
+        self.fidelity = 1.0
+        self.nodes: dict[RootPath, TreeNode] = dict()
+        self.qubit_position: dict[Qubit, tuple[RootPath, int]] = dict()
+
+        if self._cfg.truncation_fidelity < 1:
+            raise NotImplementedError(
+                "Truncation fidelity mode not currently implemented on TTN."
+            )
+
+        n_groups = len(qubit_partition)
+        if n_groups == 0:  # There's no initialisation to be done
+            return None
+        if n_groups == 1:
+            raise ValueError(
+                "Only one entry to qubit_partition provided."
+                "Introduce a finer partition of qubits."
+            )
+
+        n_levels = math.floor(math.log2(n_groups))
+        if n_groups != 2**n_levels:
+            raise ValueError(
+                "The number of entries in qubit_partition must be a power of two."
+            )
+
+        # Create the TreeNodes of the different groups of qubits
+        for k, qubits in qubit_partition.items():
+            if k < 0 or k >= n_groups:
+                raise ValueError(
+                    f"The keys of qubit_partition must range from 0 to {n_groups-1}."
+                )
+
+            # Calculate the root path of this group
+            path = []
+            for l in reversed(range(n_levels)):
+                if k < 2**l:
+                    path.append(DirTTN.LEFT)
+                else:
+                    path.append(DirTTN.RIGHT)
+                    k -= 2**l

```

Comment:
I would also need to understand this. Also, is it not dangerous to modify loop index inside the loop? Although in this case I presume not...